### PR TITLE
PolicyExtractor refactoring & testing

### PR DESCRIPTION
**Pull request by DeagleGross**
This PR is bigger than I expected due to a lot of property names fixes.

Main thing I've done:
- [x] come up with an architecture for extractor to implement in further PR's
- [x] implemented one for `PolicyExtractor`
- [x] created a test for extracting policy configuration

Future PR's will refactor other extractors also. Please, refer to linked issue for whole refactoring plan.
Relates #617

### CODE REVIEW

File: src/ArmTemplates/Extractor/Utilities/Authentication.cs [hunk 1]
```diff
@@ -9,28 +9,29 @@ namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Utilit
 {
     public class Authentication
     {
-        internal static bool isTokenValid = false;
-        public static DateTime start = DateTime.Now;
-        internal static string internalAzToken;
-        internal static string internalAzSubscriptionId;
+        internal static bool IsTokenValid = false;
+        public static DateTime Start = DateTime.Now;
+        internal static string InternalAzToken;
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Utilities/Authentication.cs:14]: Can you change these into properties?

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs [hunk 2]
```diff
@@ -1,56 +1,133 @@
...
+        public string LinkedTemplatesBaseUrl { get; private set; }
+
+        public string LinkedTemplatesSasToken { get; private set; }
+
+        public string LinkedTemplatesUrlQueryString { get; private set; }
+
+        public string PolicyXMLBaseUrl { get; private set; }
+
+        public string PolicyXMLSasToken { get; private set; }
+
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:59]: Is there a reason for having an empty line between every property (without a summary)? It just makes the file unnecessarily long and doesn't make things more readable or anything IMO .

File: src/ArmTemplates/Creator/Models/CreatorConfiguration.cs [hunk 3]
```diff
@@ -2,6 +2,7 @@
 using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.TemplateModels;
 using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
 using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Logger;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models;
 
 namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Creator.Models
```
Comment (Hardell) [inline src/ArmTemplates/Creator/Models/CreatorConfiguration.cs:7]: Can all of these classes/enums get their own file? 
Also, are they used as DTOs or why did you not apply the naming rules to this file?

File: src/ArmTemplates/Common/Templates/Abstractions/Template.cs [hunk 4]
```diff
@@ -1,16 +1,28 @@
-﻿using Newtonsoft.Json;
+﻿using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Extensions;
+using Newtonsoft.Json;
 using System.Collections.Generic;
 
 namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions
 {
     public class Template
```
Comment (Hardell) [inline src/ArmTemplates/Common/Templates/Abstractions/Template.cs:7]: This looks like a DTO. Are we not breaking anything by renaming things here?

File: src/ArmTemplates/Common/Templates/Abstractions/APITemplateResource.cs [hunk 5]
```diff
@@ -2,83 +2,119 @@
 {
     public class APITemplateResource : TemplateResource
```
Comment (Hardell) [inline src/ArmTemplates/Common/Templates/Abstractions/APITemplateResource.cs:3]: Can each class get their own file?

File: src/ArmTemplates/Common/Logger.cs [hunk 6]
```diff
@@ -9,40 +9,40 @@ namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common
...
+            InternalLogger?.LogError(message, args);
         }
 
         public static void LogError(Exception exception, string message, params object[] args)
         {
-            _logger.LogError(exception, message, args);
+            InternalLogger?.LogError(exception, message, args);
         }
     }
 }
```
Comment (Hardell) [inline src/ArmTemplates/Common/Logger.cs:48]: Isn't there a rule for having an empty line at the end of a file in the apim solutions?

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs [hunk 7]
```diff
@@ -1,56 +1,133 @@
...
+        public string ApiVersionSetName { get; private set; }
+
+        public ServiceUrlProperty[] ServiceUrlParameters { get; set; }
+
+        public bool ToParameterizeServiceUrl { get; private set; }
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:64]: Can you explain these to me please? Why do they start with a "To" and what does Parametrize mean?

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs [hunk 8]
```diff
@@ -1,56 +1,133 @@
...
+
+        public bool ToParameterizeNamedValue { get; private set; }
+
+        public bool ToParameterizeApiLoggerId { get; private set; }
+
+        public bool ToParameterizeLogResourceId { get; private set; }
+
+        public bool NotIncludeNamedValue { get; private set; }
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:72]: I would avoid "negative" bool names. It can be confusing. It's much easier to work with something that's called "ÏncludeNamedValue" instead of "NotIncludeNamedValue", especially if you need to negate it somewhere in the code(e.g. !NotIncludeNamedValue is just bad). Would it break something or be too much work to rename this one?

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs [hunk 9]
```diff
@@ -1,56 +1,133 @@
...
+        /// </summary>
+        public string SingleApiName { get; set; }
+
+        /// <summary>
+        /// Names of APIs that user wants to extract
+        /// </summary>
+        public List<string> MultipleApiNames { get; set; }
+
+        /// <summary>
+        /// Create split api templates for every API in the source APIM instance
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:41]: Do you mean separate api templates?

**Comment by Hardell** [comment]
Can we create branches in this repo instead of creating forks for PRs? It'd make switching between main and the PR branch much easier in IDEs.

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 10]
```diff
@@ -68,76 +78,52 @@ public class ExtractorExecutor
...
-                {
-                    Logger.LogInformation("Executing extraction for {0} API ...", singleApiName);
-                }
-                else
-                {
-                    Logger.LogInformation("Executing full extraction ...");
-                }
-
-                await this.GenerateTemplates(new ExtractorParameters(this.extractorConfig), singleApiName, null, fileNames, null);
+                Logger.LogInformation("No specific parameters are set, executing extraction for {0} API...", this.extractorParameters.SingleApiName);
```
Comment (Hardell) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:109]: `this.extractorParameters.SingleApiName` can be null so this log message can make no sense. Why not keep the 2 different messages depending on whether or not it's null?

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 11]
```diff
@@ -164,252 +150,267 @@ public async Task GenerateSplitAPITemplates(FileNames fileNames)
...
+                await this.GenerateTemplates(versionSetFolder, null, apiDictionary[this.extractorParameters.ApiVersionSetName], null);
+
+                Logger.LogInformation($@"Finish extracting APIVersionSet {this.extractorParameters.ApiVersionSetName}");
+            }
+        }
+
+        /// <summary>
+        /// Generates templates for multiple specified APIs
+        /// </summary>
+        public async Task GenerateMultipleAPIsTemplates()
```
Comment (Hardell) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:195]: Many of these methods can be private, right?

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 12]
```diff
@@ -164,252 +150,267 @@ public async Task GenerateSplitAPITemplates(FileNames fileNames)
...
+        {
+            // get api dictionary and check api version set
+            var apiDictionary = await this.GetAllAPIsDictionary(this.extractorParameters.SourceApimName, this.extractorParameters.ResourceGroup);
+            if (!apiDictionary.ContainsKey(this.extractorParameters.ApiVersionSetName))
+            {
+                throw new NoApiVersionSetWithSuchNameFoundException("API Version Set with this name doesn't exist");
+            }
+            else
+            {
+                Logger.LogInformation("Start extracting the API version set {0}", this.extractorParameters.ApiVersionSetName);
```
Comment (Hardell) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:173]:
```suggestion
                Logger.LogInformation("Started extracting APIVersionSet: {0}", this.extractorParameters.ApiVersionSetName);
```

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 13]
```diff
@@ -164,252 +150,267 @@ public async Task GenerateSplitAPITemplates(FileNames fileNames)
...
+        /// Generates templates for multiple specified APIs
+        /// </summary>
+        public async Task GenerateMultipleAPIsTemplates()
+        {
+            if (this.extractorParameters.MultipleApiNames.IsNullOrEmpty())
+            {
+                throw new Exception("multipleAPIs parameter doesn't have any data");
+            }
+
+            Logger.LogInformation("Start extracting multiple APIs (amount is {0} APIs)", this.extractorParameters.MultipleApiNames.Count);
```
Comment (Hardell) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:202]:
```suggestion
            Logger.LogInformation("Started extracting multiple({0}) APIs", this.extractorParameters.MultipleApiNames.Count);
```

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 14]
```diff
@@ -164,252 +150,267 @@ public async Task GenerateSplitAPITemplates(FileNames fileNames)
...
+                System.IO.Directory.CreateDirectory(apiFileFolder);
+                await this.GenerateTemplates(apiFileFolder, apiName, null, null);
+            }
+
+            // create master templates for these apis 
+            string groupApiFolder = string.Concat(this.extractorParameters.FilesGenerationRootDirectory, this.extractorParameters.FileNames.GroupAPIsMasterFolder);
+            System.IO.Directory.CreateDirectory(groupApiFolder);
+            await this.GenerateTemplates(groupApiFolder, null, this.extractorParameters.MultipleApiNames, null);
+
+            Logger.LogInformation($@"Finish extracting mutiple APIs");
```
Comment (Hardell) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:217]:
```suggestion
            Logger.LogInformation($@"Finished extracting mutiple APIs");
```

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 15]
```diff
@@ -164,252 +150,267 @@ public async Task GenerateSplitAPITemplates(FileNames fileNames)
...
+            string baseFilesGenerationDirectory,
             string singleApiName,
             List<string> multipleApiNames,
-            FileNames fileNames,
             Template apiTemplate)
         {
-            if (singleApiName != null && multipleApiNames != null)
+            if (!string.IsNullOrEmpty(singleApiName) && !multipleApiNames.IsNullOrEmpty())
             {
                 throw new SingleAndMultipleApisCanNotExistTogetherException("can't specify single API and multiple APIs to extract at the same time");
```
Comment (Hardell) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:278]:
```suggestion
                throw new SingleAndMultipleApisCanNotExistTogetherException("Can't specify single API and multiple APIs to extract at the same time");
```

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 16]
```diff
@@ -164,252 +150,267 @@ public async Task GenerateSplitAPITemplates(FileNames fileNames)
...
+                // generate seperate folder for each API
+                string apiFileFolder = string.Concat(this.extractorParameters.FilesGenerationRootDirectory, $@"/{apiName}");
+                System.IO.Directory.CreateDirectory(apiFileFolder);
+                await this.GenerateTemplates(apiFileFolder, apiName, null, null);
+            }
+
+            // create master templates for these apis 
+            string groupApiFolder = string.Concat(this.extractorParameters.FilesGenerationRootDirectory, this.extractorParameters.FileNames.GroupAPIsMasterFolder);
+            System.IO.Directory.CreateDirectory(groupApiFolder);
+            await this.GenerateTemplates(groupApiFolder, null, this.extractorParameters.MultipleApiNames, null);
```
Comment (Hardell) [inline src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:215]: There are a lot of calls all over the place where we keep passing nulls. Could those parameters be optional and null by default?

File: src/ArmTemplates/Extractor/EntityExtractors/Abstractions/EntityExtractorBase.cs [hunk 17]
```diff
@@ -18,7 +24,7 @@ public abstract class EntityExtractorBase
         static readonly IMemoryCache Cache = new MemoryCache(new MemoryCacheOptions());
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/EntityExtractors/Abstractions/EntityExtractorBase.cs:24]: Just like elsewhere. Can these be properties with private/no setters instead?

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+using System.Threading.Tasks;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Extensions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Utilities;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.API.Clients.Abstractions
+{
+    public abstract class ApiClientBase
+    {
+        public static string BaseUrl = "https://management.azure.com";
```
Comment (Hardell) [inline src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:18]: You extracted the functionality from the EntityExtractorBase, but you also duplicated this static field that is now not set anywhere anymore. So while you overwrite the default EntityExtractorBase.BaseUrl, you always use the default url in this file.

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+
+        protected Authentication Auth = new Authentication();
```
Comment (Hardell) [inline src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:20]: Just like elsewhere. Can these be properties instead?

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+
+        protected async Task<string> CallApiManagementAsync(string azToken, string requestUrl)
+        {
+            if (Cache.TryGetValue(requestUrl, out string cachedResponseBody))
+            {
+                return cachedResponseBody;
+            }
+
+            var request = new HttpRequestMessage(HttpMethod.Get, requestUrl);
+
```
Comment (Hardell) [inline src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:33]: Nit: You seem to like empty lines a lot 😃. Can you at least group some things together instead of having an empty line between almost everything? E.g. this line and the empty line below `HttpResponseMessage response` don't need to exist.

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs (hunk 18 above)
Comment (Hardell) [reply src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:18]: (related to the comment in EntityExtractorBase)
But again, can this not be a static field that's overwritten from the outside? It leads to issues exactly like this.

File: src/ArmTemplates/Common/FileHandlers/FileWriter.cs [hunk 21]
```diff
@@ -1,10 +1,37 @@
...
+            var serializedItem = item.Serialize();
+            await File.WriteAllTextAsync(Path.Combine(pathsToFile), serializedItem);
+        }
+
+        public static async Task SaveTextToFileAsync(string xmlContent, params string[] pathsToFile)
+        {
+            await File.WriteAllTextAsync(Path.Combine(pathsToFile), xmlContent);
+        }
+
+        #region Legacy writing methods
```
Comment (Hardell) [inline src/ArmTemplates/Common/FileHandlers/FileWriter.cs:33]: Why do we need to keep them at all?

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 7 above)
Comment (Hardell) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:64]: Briefly looking at the rest of the code, I still don't understand why it's called Parametrize (pls help me understand), but I think we can drop the "To".

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs [hunk 22]
```diff
@@ -1,56 +1,133 @@
...
+
+        public string ResourceGroup { get; private set; }
+
+        /// <summary>
+        /// Naming of all templates from the base folder requested by end-user in extractor-configuration parameters
+        /// </summary>
+        public FileNames FileNames { get; private set; }
+
+        /// <summary>
+        /// Not changeable root directory, where templates will be generated.
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:26]:
```suggestion
        /// The root directory where templates will be generated.
```

Why is it important to note that it's `Not changeable`? Isn't that the expected behaviour, especially since the setter is private?

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs [hunk 23]
```diff
@@ -1,56 +1,133 @@
-﻿namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.FileHandlers;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models
 {
-    public class ExtractorParameters
+    /// <summary>
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:8]: Nit: I'm not sure this summary adds any value. I'd just remove it.

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs [hunk 24]
```diff
@@ -1,56 +1,133 @@
...
+
+        public bool ToParameterizeLogResourceId { get; private set; }
+
+        public bool NotIncludeNamedValue { get; private set; }
+
+        public bool ParamNamedValuesKeyVaultSecrets { get; private set; }
+
+        public int OperationBatchSize { get; private set; }
+
+        public bool ToParameterizeBackend { get; set; }
```
Comment (Hardell) [inline src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:78]: Can you make all the props have a private setter?

File: src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs [hunk 25]
```diff
@@ -1,32 +1,87 @@
-﻿using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
+﻿using System.Collections.Generic;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Constants;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models;
 
 namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Creator.TemplateCreators
 {
-    public class TemplateCreator
+    public abstract class TemplateCreator
```
Comment (Hardell) [inline src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs:8]: TemplateCreatorBase?

Comment (tomkerkhove) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs:8]: Do we need the base suffic? 🤔

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs [hunk 26]
```diff
@@ -33,8 +35,16 @@ public class ExtractorExecutor
         readonly IApiTagExtractor apiTagExtractor;
         readonly ITagExtractor tagExtractor;
 
+        public ExtractorExecutor(ExtractorConsoleAppConfiguration extractorConfig)
+            : this(new ExtractorParameters(extractorConfig))
+        {
+            EntityExtractorBase.BaseUrl = string.IsNullOrEmpty(extractorConfig.ServiceBaseUrl)
```
[the remark this answers is not in the document]
Comment (Hardell) [reply src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:41]: There's a related comment to this in the ApiClientBase.cs

File: tests/ArmTemplates.Tests/Extractor/Scenarios/PolicyExtractorTests.cs [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+    [Trait("Category", "Policy Extraction")]
+    public class PolicyExtractorTests : ExtractorTestsBase
+    {
+        static string OutputPoliciesDirectory;
+
+        public PolicyExtractorTests()
+        {
+            OutputPoliciesDirectory = Path.Combine(TESTS_OUTPUT_DIRECTORY, "policy-tests");
+
+            // remember to clean up the output directory before each test
```
Comment (Hardell) [inline tests/ArmTemplates.Tests/Extractor/Scenarios/PolicyExtractorTests.cs:32]: Isn't this ran only once and not before each individual test?

File: tests/ArmTemplates.Tests/Moqs/ApiClients/MockPolicyApiClient.cs [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+
+        public static PolicyTemplateProperties TemplateProperties = new PolicyTemplateProperties
+        {
+            Format = "rawxml",
+            Value = @"<policies> my mocked policies </policies>"
+        };
+
+        public static IPolicyApiClient GetMockedApiClientWithDefaultValues()
+        {
+            var mockPolicyApiClient = new Mock<IPolicyApiClient>();
```
Comment (Hardell) [inline tests/ArmTemplates.Tests/Moqs/ApiClients/MockPolicyApiClient.cs:26]: Can this be strict?

File: src/ArmTemplates/Extractor/Utilities/Authentication.cs (hunk 1 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Utilities/Authentication.cs:14]: Investigated here: they are only used inside of this class, so I made that as private fields :)

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 2 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:59]: For me it is a little bit hard to read the properties characteristics (like type, visibility modifier and etc) when they are located one by one. I think it is just a personal preference.

However for that specific file you've written your comment to (`ExtractorParameters`) I maybe will write summary to every property inside, because it is like a main exposed parameters container for **Extractor**

Resolving this now, if you think it is better to be changed, feel free to say it!

File: src/ArmTemplates/Creator/Models/CreatorConfiguration.cs (hunk 3 above)
Comment (DeagleGross) [reply src/ArmTemplates/Creator/Models/CreatorConfiguration.cs:7]: I didn't refactor Creator still (it is upcoming). I dont want to change like everything right now, because I will definitely forget about something I have changed and it will be a headache later (I think we will get some bugs after refactoring is over, but hope we can reduce it's amount if we refactor things one by one).

So I would prefer to leave as it is for now.

File: src/ArmTemplates/Common/Templates/Abstractions/Template.cs (hunk 4 above)
Comment (DeagleGross) [reply src/ArmTemplates/Common/Templates/Abstractions/Template.cs:7]: I have created the default `CamelCasePropertyNamesContractResolver` to be used elsewhere
```
static readonly JsonSerializerSettings DefaultJsonSerializationSettings = new JsonSerializerSettings
        {
            ContractResolver = new CamelCasePropertyNamesContractResolver(),
            Formatting = Formatting.Indented,
            NullValueHandling = NullValueHandling.Ignore
        };
```

when the refactoring is done, every serialization\deserialiation will be working due to this code.

I decided to go with this approach instead of setting `[JsonProperty("bla-bla-bla")]` for every property. It will be quite painful :D

File: src/ArmTemplates/Common/Templates/Abstractions/APITemplateResource.cs (hunk 5 above)
Comment (DeagleGross) [reply src/ArmTemplates/Common/Templates/Abstractions/APITemplateResource.cs:3]: All classes have found their home :)

File: src/ArmTemplates/Common/Logger.cs (hunk 6 above)
Comment (DeagleGross) [reply src/ArmTemplates/Common/Logger.cs:48]: image
there is no such a setting in editor config. You want me to add it?

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 7 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:64]: It is like **ServiceUrl** is needed **To** be **Parameterized**. In another way we can say, then we (as end-users) want **To Parameterize Service Url**. (_describe or represent in terms of a parameter or parameters_)

I am trying to give a somehow understandable name to property, because before it was `ParamServiceUrl`. What is done with service url is absolutely not recognisable from such a naming. In one file I have found such a naming with fully written word **parameterize** and decided to go with that by now.

I do not understand right now what Extractor does with service url, maybe will rename it later.

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 9 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:41]: As i understand, if you are not setting `splitapi`, then all apis in APIM instance are loaded in one file. Ye, I think it is better to write `separate api templates`, thanks

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs (hunk 11 above)
Comment (DeagleGross) [reply src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:195]: Ye, made most of them private. For specific extractors methods will still be exposed so that Executor methods for specific reasons can be called from outside.

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs (hunk 12 above)
Comment (DeagleGross) [reply src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:173]: fixed

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs (hunk 14 above)
Comment (DeagleGross) [reply src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:217]: fixed

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs (hunk 13 above)
Comment (DeagleGross) [reply src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:202]: fixed

File: src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs (hunk 16 above)
Comment (DeagleGross) [reply src/ArmTemplates/Commands/Executors/ExtractorExecutor.cs:215]: Great point, fixed it

File: src/ArmTemplates/Extractor/EntityExtractors/Abstractions/EntityExtractorBase.cs (hunk 17 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/EntityExtractors/Abstractions/EntityExtractorBase.cs:24]: Fixed this in ApiClientBase. Once all extractors will migrate to this class, will delete `EntityExtractorBase` at all

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs (hunk 19 above)
Comment (DeagleGross) [reply src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:20]: changed it

File: src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs (hunk 20 above)
Comment (DeagleGross) [reply src/ArmTemplates/Common/API/Clients/Abstractions/ApiClientBase.cs:33]: I do not like them either, it was just a moving of file :)
Fixed it as I could

File: src/ArmTemplates/Common/FileHandlers/FileWriter.cs (hunk 21 above)
Comment (DeagleGross) [reply src/ArmTemplates/Common/FileHandlers/FileWriter.cs:33]: By now didn't get rid of other extractors. The same for the `EntityExtractorBase`.

I just need to make sure they all are used correctly. I am a little bit scared to change it without proper understanding how older methods are called.

Also changing like everything leads to 1000 files PR and I think it will be impossible to review

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 22 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:26]: removed that couple of words

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 23 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:8]: removed

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 24 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:78]: done

File: src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs (hunk 25 above)
Comment (DeagleGross) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs:8]: Made it static. Also changed all child-classes to use static methods. Thanks.

File: tests/ArmTemplates.Tests/Extractor/Scenarios/PolicyExtractorTests.cs (hunk 27 above)
Comment (DeagleGross) [reply tests/ArmTemplates.Tests/Extractor/Scenarios/PolicyExtractorTests.cs:32]: Extractor generates files. I decided to get such a folder structure in tests:
```
tests-output\
    policy-tests\
        <name_of_test>\
            files
            files
            files
        <name_of_test>\
            files
            files
            files
```

So it is up to a developer to clear tests output directory or not - if you are running some configuration tests, then you maybe don't want to clear all the tests-output directory (it is slowing down your tests launches).

So what is why i decided to call delete here - before each test in this class (class is for policy tests)

File: tests/ArmTemplates.Tests/Moqs/ApiClients/MockPolicyApiClient.cs (hunk 28 above)
Comment (DeagleGross) [reply tests/ArmTemplates.Tests/Moqs/ApiClients/MockPolicyApiClient.cs:26]: done

File: src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs (hunk 25 above)
Comment (Hardell) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs]: >Made it static. Also changed all child-classes to use static methods. 

Why? What was wrong with abstract? 

> Do we need the base suffic? 🤔

Not necessarily, it's just what some people prefer.

Comment (DeagleGross) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs]: I did it because of these arguments in my head:
- the class has only static methods.
- it is used not only in current child-classes. Why do we need to keep it an abstract in this case
- there are not fields or other stuff in this "abstract" class, that are used by child classes. So again no reason to leave it as abstract

File: src/ArmTemplates/Extractor/Models/ExtractorParameters.cs (hunk 8 above)
Comment (DeagleGross) [reply src/ArmTemplates/Extractor/Models/ExtractorParameters.cs:68]: Agree. Let's leave it for now, I have added a note to issue with a link to this comment so that I will not forget to do it.

File: src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs (hunk 25 above)
Comment (Hardell) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs]: How are you going to mock TemplateCreator when testing any of the classes that use it?

> Also I am not a fan of leaving a lot of methods in inheritance without a reason

If you don't want them to inherit from it, why not pass it as a ctor param instead of making it static?

> It is used in extractors also (creators + extractors both)

Can we pass it to the Extractor as a parameter then as well?

If you don't agree, please ask Tom for his opinion and let's go what what he picks.

Comment (DeagleGross) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs]: > How are you going to mock TemplateCreator when testing any of the classes that use it?
I do not want to mock this. I think mocks are needed only for some external services. But this TemplateCreator is basically a business logic of an application. We want to check this specific behaviour.

> If you don't want them to inherit from it, why not pass it as a ctor param instead of making it static?
I am not a fan of creating an instance of a class only for calling it's method. It is the reason static is better here IMO. 

> Can we pass it to the Extractor as a parameter then as well?
I am not a fan of adding a constructor for a lot of objects (you will only use it like this - `new Extractor(new TemplateCreator())`. If there is only a single way to use this class (to call it's methods), why do we need to pass it as a parameter?

In general I am not understanding the reason to leave some classes not static, when they are not having any state. I.e. we have a class Person:
```
class Person
{
   public string Name {get;set;}
}
```

This class has some state or information to hold.
Let's view TemplateCreator:
```
public static class TemplateCreator
{
    public static Template GenerateEmptyPropertyTemplateWithParameters() { ... }

    public static Template GenerateEmptyApiTemplateWithParameters(ExtractorParameters extractorParameters) { ... }

    public static Template GenerateEmptyTemplate() { ... }

    public static Template CreateEmptyParameters() { ... }
}
```

It does not hold any state or any information - it only has methods to generate new Templates.
Basically we can name it like `TemplateExtensions` or smth like this. For me `Creator` in name of class also signifies that it is more likely to be a static class, than an instantiated one.

@tomkerkhove, could you, please, kindly view this thread?

Comment (tomkerkhove) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs]: I don't really have an opinion on this but I understand both sides. However, I'm personally only using static for small utilities and not main chuncks of business logic.

Mocking is one reason, another is that you can stub them and are easier to test indeed but also extensibility. But again, I don't have a strong opinion on it.

Comment (Hardell) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs]: > I do not want to mock this. I think mocks are needed only for some external services. But this TemplateCreator is basically a business logic of an application. We want to check this specific behaviour.

Yes, and we would check **once** it because it's be its own class that we would have UTs for. If it's static, you need to test it in every single class where you use it, because you can't mock it. And if you use a static class within a static class that's then used in a class that you want to test, you'll probably have to test things that are only remotely related to the class you're testing.

> I am not a fan of creating an instance of a class only for calling it's method. It is the reason static is better here IMO.

My arguments from above apply here. You can't properly test static things.

> I am not a fan of adding a constructor for a lot of objects (you will only use it like this - new Extractor(new TemplateCreator()). 

Ideally we'd want to have autofac set up. You can then create a TemplateCreator as a single instance because it's stateless and then pass it everywhere as a ctor param.

> If there is only a single way to use this class (to call it's methods), why do we need to pass it as a parameter?

So that you can test it :) I feel like this entire discussion comes down to the fact that you cannot properly test things you can't mock and that you can't mock static calls.

> In general I am not understanding the reason to leave some classes not static, when they are not having any state

Same as above - because you want to be able to test it and mock it. It's the basic concept of dependency injection that you want to make classes independent from its dependencies. Without it, you do a change a single static class and now you have to fix 500 tests that have nothing to do with the static class because other classes that used the static class were not independent from it.

Comment (DeagleGross) [reply src/ArmTemplates/Creator/TemplateCreators/TemplateCreator.cs]: Pushed changes, made it using inheritance. Thanks for your opinions and help!)

**Review by Hardell — APPROVED** [review]
